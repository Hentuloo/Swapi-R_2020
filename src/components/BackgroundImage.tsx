import React, { FC } from 'react';
import styled from 'styled-components';
import { background } from 'assets/images/backgrounds';

const Wrapper = styled.img`
    position: fixed;
    min-height: 60%;
    min-width: 60%;
    max-width: 1300px;
    left: 50%;
    top: 50%;
    z-index: -1;
    transform: translate(-50%, -50%);
    opacity: 0.15;
`;

export interface BackgroundImageProps {}

export const BackgroundImage: FC<BackgroundImageProps> = ({ ...props }) => {
    return (
        <Wrapper
            {...props}
            sizes="(max-width: 2240px) 100vw, 2240px"
            srcSet={`
                    ${background[0]} 1024w,
                    ${background[1]} 1539w,
                    ${background[2]} 1891w,
                    ${background[3]} 2171w,
                    ${background[4]} 2240w`}
            src={background[4]}
        />
    );
};
