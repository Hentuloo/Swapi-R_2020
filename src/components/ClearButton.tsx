import styled from 'styled-components';

export const ClearButton = styled.button`
    border: none;
    background-color: transparent;
    color: ${({ theme }) => theme.color.white[0]};
    text-transform: uppercase;
    text-decoration: underline;
    padding: 10px 10px;
`;
