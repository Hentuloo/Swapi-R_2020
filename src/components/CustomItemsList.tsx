import React, { FC } from 'react';
import styled from 'styled-components';
import { LabelsListProps, LabelsList, LabelWrapper } from './LabelsList';

const Wrapper = styled.div`
    width: 80%;
    max-width: 800px;
    height: 100%;
    margin: 0px auto;
    ${({ theme }) => theme.mediaQuery.lg} {
        width: 50%;
        max-width: 1200px;
        margin: 0px 0px 0px auto;
    }
`;
const StyledLabelsList = styled(LabelsList)`
    display: grid;
    width: 90%;
    grid-row-gap: 30px;
    grid-column-gap: 10px;
    grid-template-columns: 1fr 1fr;
    margin: 70px auto 0px;
    padding-bottom: 100px;
    justify-items: center;

    ${({ theme }) => theme.mediaQuery.lg} {
        width: 100%;
        max-width: none;
    }
    ${LabelWrapper} {
        max-width: 150px;
        ${({ theme }) => theme.mediaQuery.vlg} {
            max-width: 200px;
        }
    }
    ${LabelWrapper}:nth-child(odd) {
        margin-top: 40px;
    }
    ${LabelWrapper}:nth-child(even) {
        margin-top: -40px;
    }
`;

export interface CustomItemsListProps extends LabelsListProps {}

export const CustomItemsList: FC<CustomItemsListProps> = ({
    items,
    ...props
}) => {
    return (
        <Wrapper {...props}>
            <StyledLabelsList items={items} />
        </Wrapper>
    );
};
