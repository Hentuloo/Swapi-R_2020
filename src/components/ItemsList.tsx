import React, { FC } from 'react';
import styled from 'styled-components';
import { LabelsListProps, LabelsList } from './LabelsList';

const Wrapper = styled.div`
    width: 80%;
    max-width: 800px;
    height: 100%;
    margin: 0px auto;
    ${({ theme }) => theme.mediaQuery.lg} {
        width: 90%;
        max-width: 1200px;
    }
`;
const StyledLabelsList = styled(LabelsList)`
    ${({ theme }) => theme.mediaQuery.lg} {
        max-width: auto;
        height: 100%;
        display: grid;
        grid-template-columns: 1fr 1fr;
        grid-column-gap: 150px;
    }
`;

export interface ItemsListProps extends LabelsListProps {}

export const ItemsList: FC<ItemsListProps> = ({ items, ...props }) => {
    return (
        <Wrapper {...props}>
            <StyledLabelsList items={items} />
        </Wrapper>
    );
};
