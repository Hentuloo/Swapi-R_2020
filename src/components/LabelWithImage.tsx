import React, { FC } from 'react';
import styled, { css } from 'styled-components';
import { Link } from 'react-router-dom';
import { useImage } from 'react-image';
import defaultCharacterImage from 'assets/images/defaultCharacter.svg';

export type LabelMode = 'SQUARE' | 'CIRCLE';

interface ComponentModeProps {
    to?: string;
    mode?: LabelMode;
}
export const Wrapper = styled.div<ComponentModeProps>`
    position: relative;
    display: block;
    font-size: ${({ theme }) => theme.fs.l};
    color: ${({ theme }) => theme.color.white[0]};
    text-decoration: none;
    border-radius: 31px;
    overflow: hidden;
    ${({ mode }) =>
        mode === 'CIRCLE' &&
        css`
            display: grid;
            grid-template-columns: 70px 1fr;
            align-items: center;
            grid-column-gap: 20px;
        `}
`;

export const ContentWrapper = styled.div<ComponentModeProps>`
    position: absolute;
    bottom: 15%;
    left: 5%;
    width: 90%;
    ${({ mode }) =>
        mode === 'CIRCLE' &&
        css`
            position: relative;
            bottom: auto;
            left: auto;
        `}
`;
export const ImageWrapper = styled.div<ComponentModeProps>`
    width: 100%;
    ${({ mode }) =>
        mode === 'CIRCLE' &&
        css`
            height: 70px;
            border-radius: 50%;
            overflow: hidden;
        `}
`;
export const Image = styled.img`
    display: block;
    width: 100%;
`;

export interface LabelWithImageProps {
    src?: string;
    title?: string;
    alt?: string;
    to?: string;
    mode?: LabelMode;
    defaultImage?: string;
    suspense?: boolean;
}

export const LabelWithImage: FC<LabelWithImageProps> = ({
    children,
    src = defaultCharacterImage,
    title,
    alt,
    to,
    defaultImage,
    mode = 'SQUARE',
    suspense = false,
    ...props
}) => {
    const { src: suspendedSrc } = useImage({
        srcList: src,
        useSuspense: suspense,
    });

    const handleError = ({
        currentTarget,
    }: React.SyntheticEvent<HTMLImageElement, Event>) => {
        if (!defaultImage) return;
        currentTarget.src = defaultImage || '';
    };

    return (
        <Wrapper
            {...props}
            as={to ? Link : 'div'}
            to={to}
            title={title}
            mode={mode}
        >
            <ImageWrapper mode={mode}>
                <Image src={suspendedSrc} alt={alt} onError={handleError} />
            </ImageWrapper>
            <ContentWrapper mode={mode}>{children}</ContentWrapper>
        </Wrapper>
    );
};
