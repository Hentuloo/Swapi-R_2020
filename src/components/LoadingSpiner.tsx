import React, { FC } from 'react';
import styled, { keyframes } from 'styled-components';
import loadingSVG from 'assets/images/loading.svg';

const rotate = keyframes`
  from {
    transform: rotate(0deg);
  }

  to {
    transform: rotate(-360deg);
  }
`;

const Wrapper = styled.div`
    margin: 20px;
`;
const SpinerImage = styled.img`
    animation: ${rotate} 2s linear infinite;
`;

export interface LoadingSpinerProps {}

export const LoadingSpiner: FC<LoadingSpinerProps> = ({ ...props }) => {
    return (
        <Wrapper {...props}>
            <SpinerImage src={loadingSVG} />
        </Wrapper>
    );
};
