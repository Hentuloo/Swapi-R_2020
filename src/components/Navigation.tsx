import React, { FC, memo } from 'react';
import styled from 'styled-components';
import { Link } from 'react-router-dom';
import { Routes } from 'config/Routes';
import deathStar from 'assets/images/deathStar.svg';

const Wrapper = styled.nav`
    position: fixed;
    left: 0%;
    bottom: 0%;
    width: 100%;
    color: ${({ theme }) => theme.color.brand[0]};
    font-size: ${({ theme }) => theme.fs.l};
    ${({ theme }) => theme.mediaQuery.md} {
        width: 250px;
        bottom: auto;
        top: 50%;
        font-size: ${({ theme }) => theme.fs.xl};
    }
`;
const ListWrapper = styled.ul`
    display: grid;
    grid-auto-flow: column;
    ${({ theme }) => theme.mediaQuery.md} {
        grid-auto-flow: row;
        top: 50%;
        transform: translate(0%, -50%);
    }
`;
const ListElement = styled.li`
    list-style: none;
`;
const StyledLink = styled(Link)`
    display: block;
    padding: 25px 0px;
    text-align: center;
    text-decoration: none;
    color: inherit;
    ${({ theme }) => theme.mediaQuery.md} {
        padding: 35px 0px;
    }
`;
const BackgroundImage = styled.img`
    display: none;
    position: fixed;
    z-index: -1;
    ${({ theme }) => theme.mediaQuery.md} {
        display: block;
        height: 100%;
        left: 0%;
        bottom: 50%;
        transform: translate(-66%, 45%) rotate(15deg);
    }
`;

export interface NavigationProps {}

export const Navigation: FC<NavigationProps> = ({ ...props }) => {
    return (
        <Wrapper {...props}>
            <ListWrapper>
                <ListElement>
                    <StyledLink to={Routes.PATHS.characters.path}>
                        {Routes.PATHS.characters.title}
                    </StyledLink>
                </ListElement>
                <ListElement>
                    <StyledLink to={Routes.PATHS.vehicles.path}>
                        {Routes.PATHS.vehicles.title}
                    </StyledLink>
                </ListElement>
                <ListElement>
                    <StyledLink to={Routes.PATHS.planets.path}>
                        {Routes.PATHS.planets.title}
                    </StyledLink>
                </ListElement>
            </ListWrapper>
            <BackgroundImage src={deathStar} title="Gwiazda Å›mierci" />
        </Wrapper>
    );
};
export const MemoizedNavigation = memo(Navigation);
