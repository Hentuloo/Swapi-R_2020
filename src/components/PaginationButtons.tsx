import React, { FC } from 'react';
import styled from 'styled-components';
import { ClearButton } from './ClearButton';

const Wrapper = styled.div``;

export const CenteredText = styled.span``;

export interface PaginationButtonsProps {
    prev?: () => void;
    next?: () => void;
    active?: number;
    maxPage?: number;
}

export const PaginationButtons: FC<PaginationButtonsProps> = ({
    prev,
    next,
    active,
    maxPage,
    ...props
}) => {
    return (
        <Wrapper {...props}>
            {prev && <ClearButton onClick={prev}>Prev</ClearButton>}
            {active && (
                <CenteredText>
                    {active}
                    {maxPage && `/${maxPage}`}
                </CenteredText>
            )}
            {next && <ClearButton onClick={next}>next</ClearButton>}
        </Wrapper>
    );
};
