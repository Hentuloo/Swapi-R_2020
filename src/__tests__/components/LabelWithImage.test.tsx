import React from 'react';
import { render, wait } from 'tests/render';
import { LabelWithImage } from 'components/LabelWithImage';

const imageElementId = 'LabelWithImage/image';

describe('LabelWithImage component', () => {
    it('snapshot', () => {
        const { container } = render(<LabelWithImage />);
        expect(container.firstChild).toMatchSnapshot();
    });
    it('should render children components', async () => {
        const text = 'Hello world';
        const { findByText } = render(<LabelWithImage>{text}</LabelWithImage>);

        const el = await findByText(text);
        expect(el).toBeInTheDocument();
    });
    it('should render as div in default', () => {
        const { container } = render(<LabelWithImage />);

        const wrapper = container.firstChild;
        const wrapperNodeName = wrapper && wrapper.nodeName;

        expect(wrapperNodeName).toBe('DIV');
    });
    it('should render as Link', () => {
        const { container } = render(<LabelWithImage to="/" />);

        const wrapper = container.firstChild;
        const wrapperNodeName = wrapper && wrapper.nodeName;

        expect(wrapperNodeName).toBe('A');
    });
});
