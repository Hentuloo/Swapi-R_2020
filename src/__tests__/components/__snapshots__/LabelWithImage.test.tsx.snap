// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LabelWithImage component snapshot 1`] = `
<div
  class="sc-AxjAm ZveYB"
  mode="SQUARE"
>
  <div
    class="sc-AxiKw cYfdGU"
    mode="SQUARE"
  >
    <img
      class="sc-AxhCb cwNtVx"
    />
  </div>
  <div
    class="sc-AxirZ ULhtM"
    mode="SQUARE"
  />
</div>
`;
